processor/intracesampler/processor.go

```diff
@@ -0,0 +1,224 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intracesampler // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/intracesamplerprocessor"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/processor"
+	"go.opentelemetry.io/collector/processor/processorhelper"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	// The constants help translate user friendly percentages to numbers direct used in sampling.
+	numHashBuckets        = 0x4000 // Using a power of 2 to avoid division.
+	bitMaskHashBuckets    = numHashBuckets - 1
+	percentageScaleFactor = numHashBuckets / 100.0
+)
+
+type inTraceSamplerProcessor struct {
+	logger             *zap.Logger
+	config             Config
+	scaledSamplingRate uint32
+	hashSeedBytes      []byte
+}
+
+func newInTraceSamplerSpansProcessor(ctx context.Context, set processor.CreateSettings, cfg *Config, nextConsumer consumer.Traces) (processor.Traces, error) {
+
+	its := &inTraceSamplerProcessor{
+		logger:             set.Logger,
+		config:             *cfg,
+		scaledSamplingRate: uint32(cfg.SamplingPercentage * percentageScaleFactor),
+		hashSeedBytes:      i32tob(cfg.HashSeed),
+	}
+
+	return processorhelper.NewTracesProcessor(
+		ctx,
+		set,
+		cfg,
+		nextConsumer,
+		its.processTraces,
+		processorhelper.WithCapabilities(consumer.Capabilities{MutatesData: true}))
+}
+
+type FullSpan struct {
+	resource pcommon.Resource
+	scope    pcommon.InstrumentationScope
+	span     ptrace.Span
+}
+
+type TraceTreeData struct {
+
+	// map each span id to a full span object with scope and resource
+	fullSpans map[pcommon.SpanID]FullSpan
+
+	// map each span id to its children ids
+	// this enables fast leaf detection and traversal of the tree from roots
+	children map[pcommon.SpanID][]pcommon.SpanID
+
+	// spans that have no parent or their parent is not in the trace
+	roots []pcommon.SpanID
+}
+
+// this map enable us to find the parent span of a span in O(1), and all the children of a givin span
+// it also generates a span object that contains the resource and scope all at once
+// this is useful for the sampler to be able to make decisions on spans
+func spansToTraceTree(td ptrace.Traces) TraceTreeData {
+	fullSpans := make(map[pcommon.SpanID]FullSpan)
+	spanChildren := make(map[pcommon.SpanID][]pcommon.SpanID)
+
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		resource := rs.Resource()
+		scopespans := rs.ScopeSpans()
+		for j := 0; j < scopespans.Len(); j++ {
+			ss := scopespans.At(j)
+			spans := ss.Spans()
+			scope := ss.Scope()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				fullspan := FullSpan{
+					resource: resource,
+					scope:    scope,
+					span:     span,
+				}
```

Comment:
In order to reduce on the number of required allocation, you can make this:

```golang
fullspan := FullSpan{
    resource : rs.Resource(),
    scope     : ss.Scope(),
    span       : span, 
}
```

Reply:
no problem, but isn't it cheaper to cache the getter function result once and use it for all iterations of the loop? instead of invoking it again and again for the nested loops?